### ISSUE TITLE
[2013-01-21] getFinalisedQuery misses tables when looking for columns

### ISSUE THREAD

**Issue by silverstripe-issues** [silverstripe/silverstripe-framework#1669]
created by: weboutreach
created at: 2013-01-21
original ticket: http://open.silverstripe.org/ticket/8208

---

The regular expression that extracts the column name from expressions like '"Table"."Column"' returns the table rather than the column.

**Comment by ss23** [comment]
Is this really critical if all tests are passing? I'm a little confused as to how things are working currently.

**Comment by dhensby** [comment]
I assume there's a test somewhere that proves this fails. However I recall this issue coming up more recently and think it may have been patched...

**Comment by ss23** [comment]
@dhensby Sounds like we should close this ticket then. You have that permission, right?

**Comment by micmania1** [comment]
Not sure if this is the same issue: https://github.com/silverstripe/silverstripe-framework/issues/1682

**Comment by dhensby** [comment]
That's not the same issue, but I think this was caught when that was patched. 

I can close this, but don't want to unless I am sure it is fixed.

**Comment by ss23** [comment]
@dhensby I don't want to keep it open unless I'm sure it's still an issue :P

**Comment by kinglozzer** [comment]
Can we close this @dhensby @tractorcow?

**Comment by tractorcow** [comment]
Sorry, just checked this code... it is an issue.

Existing code:

```php
if (preg_match('/^"([^"]+)"/', $where, $matches) ||
	preg_match('/^"([^"]+)"\."[^"]+"/', $where, $matches)) {
	if (!in_array($matches[1], $queriedColumns)) $queriedColumns[] = $matches[1];
}
```

Probably should be:

```php
if (preg_match('/^"([^"]+)"/', $where, $matches) ||
	preg_match('/^"[^"]+"\."([^"]+)"/', $where, $matches)) {
	if (!in_array($matches[1], $queriedColumns)) $queriedColumns[] = $matches[1];
}
```

All I did was move the () across a bit.

A risk here is that $where could actually contain multiple columns, but we only check the first matched one.

I might be able to spend a bit of time looking at this. :D Put it in a separate method with unit tests, maybe.

**Comment by tractorcow** [comment]
I'll aim to fix it in the 3 branch, that way I can add new api for unit testing.

**Comment by dhensby** [comment]
:)

**Comment by tractorcow** [comment]
I started writing up a solution. New method candidate:


```php

class SQLConditionalExpression {
	public function getWhereColumns() {
		$columns = array();
		foreach ($this->getWhereParameterised($parameters) as $where) {
			// Check all the columns in the form '"Column" = ?' and the form '"Table"."Column"' = ?
			if(preg_match_all('/("(?<table>[^"]+)"\.)?"(?<column>[^"])+"([^\.]|$)/', $where, $matches)) {
				foreach($matches as $match) {
					$columns[$match['column']] = isset($matches['table']) ? $matches['table'] ? true;
				}
			}
		}
		return $columns;
	}
```

**Comment by tractorcow** [comment]
I've just about finished writing up the tests. See https://github.com/tractorcow/silverstripe-framework/commits/pulls/3.3/fix-finalised-query

**Comment by dhensby** [comment]
Beaty

**Comment by tractorcow** [comment]
Fix at https://github.com/silverstripe/silverstripe-framework/pull/5579

**Comment by tractorcow** [comment]
Fixed with #5579
